authz/providers/azure/rbac/checkaccessreqhelper.go

```diff
@@ -0,0 +1,338 @@
+	ActionId            string              `json:"actionId"`
+	IsDataAction        bool                `json:"isDataAction"`
+	AzureRoleAssignment AzureRoleAssignment `json:"roleAssignment"`
+	AzureDenyAssignment AzureDenyAssignment `json:"denyAssignment"`
+	TimeToLiveInMs      int                 `json:"timeToLiveInMs"`
+}
+
+func getScope(resourceId string, attr *authzv1.ResourceAttributes) string {
+	if attr != nil && attr.Namespace != "" {
+		return path.Join(resourceId, namespaces, attr.Namespace)
+	}
+	return resourceId
+}
+
+func isValidUUID(u string) bool {
+	_, err := uuid.Parse(u)
+	return err == nil
+}
+
+func getValidSecurityGroups(groups []string) []string {
+	var finalGroups []string
+	for _, element := range groups {
+		if isValidUUID(element) {
+			finalGroups = append(finalGroups, element)
+		}
+	}
+	return finalGroups
+}
+
+func getActionName(verb string) string {
+	/* special verbs
+	use verb on podsecuritypolicies resources in the policy API group
+	bind and escalate verbs on roles and clusterroles resources in the rbac.authorization.k8s.io API group
+	impersonate verb on users, groups, and serviceaccounts in the core API group
+	userextras in the authentication.k8s.io API group
+
+	https://kubernetes.io/docs/reference/access-authn-authz/authorization/#determine-the-request-verb
+	*/
+	switch verb {
+	case "get":
+		fallthrough
+	case "list":
+		fallthrough
+	case "watch":
+		return "read"
+
+	case "bind":
+		return "bind/action"
+	case "escalate":
+		return "escalate/action"
+	case "use":
+		return "use/action"
+	case "impersonate":
+		return "impersonate/action"
+
+	case "create":
+		fallthrough //instead of action create will be mapped to write
+	case "patch":
+		fallthrough
+	case "update":
+		return "write"
+
+	case "delete":
+		fallthrough
+	case "deletecollection": // TODO: verify scenario
+		return "delete"
+	default:
+		return ""
+	}
+}
+
+func getDataAction(subRevReq *authzv1.SubjectAccessReviewSpec, clusterType string) AuthorizationActionInfo {
+	authInfo := AuthorizationActionInfo{
+		IsDataAction: true}
+
+	authInfo.AuthorizationEntity.Id = clusterType
+	if subRevReq.ResourceAttributes != nil {
+		if subRevReq.ResourceAttributes.Group != "" {
+			authInfo.AuthorizationEntity.Id = path.Join(authInfo.AuthorizationEntity.Id, subRevReq.ResourceAttributes.Group)
+		}
+		authInfo.AuthorizationEntity.Id = path.Join(authInfo.AuthorizationEntity.Id, subRevReq.ResourceAttributes.Resource, getActionName(subRevReq.ResourceAttributes.Verb))
+	} else if subRevReq.NonResourceAttributes != nil {
+		authInfo.AuthorizationEntity.Id = path.Join(authInfo.AuthorizationEntity.Id, subRevReq.NonResourceAttributes.Path, getActionName(subRevReq.NonResourceAttributes.Verb))
+	}
+	return authInfo
+}
+
+func getResultCacheKey(subRevReq *authzv1.SubjectAccessReviewSpec) string {
+	cacheKey := subRevReq.User
+
+	if subRevReq.ResourceAttributes != nil {
+		if subRevReq.ResourceAttributes.Namespace != "" {
```

Comment:
I think this can be tricked by creating a namespace that matches a group. We should always set a path component even for empty ns or group.

```
	cacheKey = path.Join(cacheKey, defaultDir(subRevReq.ResourceAttributes.Namespace))
	cacheKey = path.Join(cacheKey, defaultDir(subRevReq.ResourceAttributes.Group))


func defaultDir(s string) string {
	if s != "" {
		return s
	}
	return "-" // invalid for a namespace
}
```